nomt/src/beatree/ops/update/branch_updater.rs

```diff
@@ -137,66 +183,147 @@ impl BranchUpdater {
         self.cutoff = None;
     }
 
-    fn keep_up_to(&mut self, up_to: Option<&Key>) {
-        while let Some(next_key) = self.base.as_ref().and_then(|b| b.next_key()) {
-            let Some(ref mut base_node) = self.base else {
-                return;
+    // Advance the base looking for `up_to`, stops if a bigger Key is found or the end is reached.
+    // Collect in `self.ops` all separators that are skipped.
+    // Returns the index at which 'up_to' was found, otherwise, returns None.
+    fn keep_up_to(&mut self, up_to: Option<&Key>) -> Option<usize> {
+        if self.base.is_none() {
+            // empty db
+            return None;
+        }
+
+        let (from, base_compressed_end, to, found) = {
+            // UNWRAP: self.base is not None
+            let base = self.base.as_mut().unwrap();
+
+            let from = base.low;
+            let base_n = base.node.n() as usize;
+
+            let (found, to) = match up_to {
+                // Nothing more to do, the end has already been reached
+                None if from == base_n => return None,
+                // Jump directly to the end of the base node and update `base.low` accordingly
+                None => {
+                    base.low = base_n;
+                    (false, base_n)
+                }
+                Some(up_to) => match base.find_key(up_to) {
+                    Some(res) => res,
+                    // already at the end
+                    None => return None,
+                },
             };
 
-            let order = up_to
-                .map(|up_to| up_to.cmp(&next_key))
-                .unwrap_or(Ordering::Greater);
-            if order == Ordering::Less {
-                break;
+            if from == to {
+                // nothing to keep
+                return None;
             }
 
-            if order == Ordering::Greater {
-                let separator_len = separator_len(&next_key);
-                self.ops
-                    .push(BranchOp::Keep(base_node.iter_pos, separator_len));
+            let base_compressed_end = std::cmp::min(to, base.node.prefix_compressed() as usize);
 
-                base_node.advance_iter();
-                self.bulk_split_step(self.ops.len() - 1);
-            } else {
-                base_node.advance_iter();
-                break;
-            }
+            (from, base_compressed_end, to, found)
```

Comment:
this permits `from == base_compressed_end`, leading to us pushing an empty chunk.

Reply:
Absolutely yes, thanks for funding out the bug! You have already addressed the problem in #597!